modules/ws_bridge/src/ipc_graph.cpp

```diff
@@ -0,0 +1,209 @@
+  for (const auto& [topic, type] : state_.topics_to_types_map) {
+    if (topic.length() > max_topic_length) {
+      max_topic_length = topic.length();
+    }
+    if (type.length() > max_type_length) {
+      max_type_length = type.length();
+    }
+  }
+
+  // Create the formatted string
+  for (const auto& [topic, type] : state_.topics_to_types_map) {
+    result << " - " << std::left << std::setw(static_cast<int>(max_topic_length)) << topic
+           << "\tType: " << std::left << std::setw(static_cast<int>(max_type_length)) << type << "\n";
+  }
+
+  return result.str();
+}
+
+TopicsToTypesMap IpcGraph::getTopicsToTypesMap() const {
+  absl::MutexLock lock(&mutex_);
+  return state_.topics_to_types_map;
+}
+
+TopicsToTypesMap IpcGraph::getServicesToTypesMap() const {
+  absl::MutexLock lock(&mutex_);
+  return state_.services_to_types_map;
+}
+
+TopicsToTypesMap IpcGraph::getServicesToNodesMap() const {
+  absl::MutexLock lock(&mutex_);
+  return state_.services_to_nodes_map;
+}
+
+TopicToNodesMap IpcGraph::getTopicToSubscribersMap() const {
+  absl::MutexLock lock(&mutex_);
+  return state_.topic_to_subscribers_map;
+}
+
+TopicToNodesMap IpcGraph::getTopicToPublishersMap() const {
+  absl::MutexLock lock(&mutex_);
+  return state_.topic_to_publishers_map;
+}
+
+void IpcGraph::addPublisher(const ipc::zenoh::EndpointInfo& info) {
+  absl::MutexLock lock(&mutex_);
+  if (!addTopic(info.topic)) {
+    return;
+  }
+
+  state_.topic_to_publishers_map[info.topic].push_back(info.session_id);
+}
+
+void IpcGraph::removePublisher(const ipc::zenoh::EndpointInfo& info) {
+  absl::MutexLock lock(&mutex_);
+  auto& publishers = state_.topic_to_publishers_map[info.topic];
+  publishers.erase(std::remove(publishers.begin(), publishers.end(), info.session_id), publishers.end());
+  if (publishers.empty()) {
+    state_.topic_to_publishers_map.erase(info.topic);
+    removeTopic(info.topic);
+  }
+}
+
+bool IpcGraph::hasPublisher(const std::string& topic) {
+  absl::MutexLock lock(&mutex_);
+  return state_.topic_to_publishers_map.find(topic) != state_.topic_to_publishers_map.end();
+}
+
+void IpcGraph::addSubscriber(const ipc::zenoh::EndpointInfo& info) {
+  absl::MutexLock lock(&mutex_);
+  state_.topic_to_subscribers_map[info.topic].push_back(info.session_id);
+}
+
+void IpcGraph::removeSubscriber(const ipc::zenoh::EndpointInfo& info) {
+  absl::MutexLock lock(&mutex_);
+  auto& subscribers = state_.topic_to_subscribers_map[info.topic];
+  subscribers.erase(std::remove(subscribers.begin(), subscribers.end(), info.session_id), subscribers.end());
+  if (subscribers.empty()) {
+    state_.topic_to_subscribers_map.erase(info.topic);
+  }
+}
+
+void IpcGraph::removeTopic(const std::string& topic) {
+  absl::MutexLock lock(&mutex_);
+  state_.topics_to_types_map.erase(topic);
+  state_.topic_to_publishers_map.erase(topic);
+  state_.topic_to_subscribers_map.erase(topic);
+
+  heph::log(heph::INFO, "[IPC Graph] - Topic dropped: '", topic, "'");
```

Comment:
For the logging we use structured logging so the format is:
* message
* list of pairs: key(str), value (any formattable type)
```suggestion
  heph::log(heph::INFO, "[IPC Graph] - Topic dropped", "topic", topic);
```